setup.pl

```diff
@@ -533,23 +537,132 @@
 		}
 	close(START);
-	$start_cmd = "$config_directory/start";
+	$start_cmd = "$config_directory/start-init";
 
-	open(STOP, ">$config_directory/stop");
+	# Stop main
+	open(STOP, ">$config_directory/stop-init");
 	print STOP "#!/bin/sh\n";
-	print STOP "echo Stopping Webmin server in $wadir\n";
+	print STOP "if [ \"\$1\" = \"--kill\" ]; then\n";
+	print STOP "  echo Force stopping Webmin server in $wadir\n";
+	print STOP "else\n";
+	print STOP "  echo Stopping Webmin server in $wadir\n";
+	print STOP "fi\n";
 	print STOP "pidfile=\`grep \"^pidfile=\" $config_directory/miniserv.conf | sed -e 's/pidfile=//g'\`\n";
-	print STOP "kill \`cat \$pidfile\`\n";
+	print STOP "pid=\`cat \$pidfile\` 2>/dev/null\n";
+	print STOP "if [ \"\$pid\" != \"\" ]; then\n";
+	print STOP "  kill \$pid || exit 1\n";
+	print STOP "  touch $var_dir/stop-flag\n";
+	print STOP "  if [ \"\$1\" = \"--kill\" ]; then\n";
+	print STOP "    sleep 1\n";
+	print STOP "    ((ps axf | grep \"webmin\\\/miniserv\\.pl\" | awk '{print \"kill -9 -- -\" \$1}' | bash) || kill -9 -- -\$pid || kill -9 \$pid) 2>/dev/null\n";
+	print STOP "  fi\n";
+	print STOP "  exit 0\n";
+	print STOP "else\n";
+	print STOP "  if [ \"\$1\" = \"--kill\" ]; then\n";
+	print STOP "    (ps axf | grep \"webmin\\\/miniserv\\.pl\" | awk '{print \"kill -9 -- -\" \$1}' | bash) 2>/dev/null\n";
+	print STOP "  fi\n";
+	print STOP "fi\n";
 	close(STOP);
 
-	open(RESTART, ">$config_directory/restart");
+	# Restart main
+	open(RESTART, ">$config_directory/restart-init");
 	print RESTART "#!/bin/sh\n";
-	print RESTART "$config_directory/stop && $config_directory/start\n";
+	print RESTART "$config_directory/stop-init\n";
+	print RESTART "$config_directory/start-init\n";
 	close(RESTART);
-
-	chmod(0755, "$config_directory/start", "$config_directory/stop",
-		    "$config_directory/restart");
+	
+	# Force reload main
+	open(FRELOAD, ">$config_directory/force-reload-init");
+	print FRELOAD "#!/bin/sh\n";
+	print FRELOAD "$config_directory/stop-init --kill\n";
+	print FRELOAD "$config_directory/start-init\n";
+	close(FRELOAD);
+
+	# Reload main
+	open(RELOAD, ">$config_directory/reload-init");
+	print RELOAD "#!/bin/sh\n";
+	print RELOAD "echo Reloading Webmin server in $wadir\n";
+	print RELOAD "pidfile=\`grep \"^pidfile=\" $config_directory/miniserv.conf | sed -e 's/pidfile=//g'\`\n";
+	print RELOAD "kill -USR1 \`cat \$pidfile\`\n";
+	close(RELOAD);
+
+	chmod(0755, "$config_directory/start-init");
+	chmod(0755, "$config_directory/stop-init");
+	chmod(0755, "$config_directory/restart-init");
+	chmod(0755, "$config_directory/force-reload-init");
+	chmod(0755, "$config_directory/reload-init");
+
+	# Re-generating supplementary
+
+	# Clear existing
+	my $csupp = sub {
+		unlink("$config_directory/start");
+		unlink("$config_directory/stop");
+		unlink("$config_directory/restart");
+		unlink("$config_directory/force-reload");
+		unlink("$config_directory/reload");
+	};
+	&$csupp();
```

Comment:
Why put this into a sub instead of just executing the statements?

Reply:
Because there are two places where you need to call exact the same code. If we ever again change this code we would need to remember to change it two times, in this case only once.

Reply:
AH, ok .. makes sense.